### ISSUE TITLE
Rule idea: disallow THIS-OBJECT based references to object instance members

### ISSUE THREAD

**Issue by mikefechner** [Riverside-Software/sonar-openedge#1022]
According to 

https://community.progress.com/s/article/4GL-ABL-Different-performance-when-using-not-using-this-object?popup=true

and Peter's tests the usage of THIS-OBJECT for referencing properties is 25% slower than using the unqualified reference to the instance member. 

This would not be as trivial as to ban the THIS-OBJECT keyword, as it's still required in other cases:

- Constructor chaining
- passing a reference as a parameter to an external method call
- ...

**Comment by gquerret** [comment]
This is the kind of rule that makes me wonder what the compiler does exactly !
But also the kind of rule that should be aware of the version which is being used in the project, as it won't make any sense once the bug is fixed. If it's not in the default quality profile, many people will never see it, and I don't want it to raise lots of false positives for something that is not a bug anymore in the latest versions.

**Comment by mikefechner** [comment]
That is a luxury problem to have ... For now the bug persists.

**Comment by gquerret** [comment]
@mikefechner I didn't have time this week to have a look at the KB entry or to test it. If we're only talking about 300ms overhead for 10m iterations, then there's really no need for a rule.

**Comment by stefandrissen** [comment]
Example on [abldojo](https://abldojo.services.progress.com/?shareId=63d1a3f1d172aa40d044dd8a).

````
OpenEdge V. 12.6.0
Platform: Linux (x86_64)

iterations  10000000 
loop        846 ms 

this-object 2626 ms 
no-object   2267 ms
````

**Comment by ccecvb** [comment]
I agree, the cost of changing all sources to comply with the new rule is probably higher than the possible gain.
IMHO, it's yet another example of the OpenEdge compiler not doing any optimisation. It should be fixed by Progress as this kind of issue makes our developers frown their eyebrows in disbelief.
